### PR TITLE
Resolve StepDirection invalid issue: move indicator to spectrum_loss

### PR DESCRIPTION
**Pull request by jibanCat**
The problem of invalid step size is because there is a line calculating the gradient of log(beta) like this:

```matlab
 % gradient wrt log β
  da = da .* log(lya_1pz) * beta;
  dlog_beta  = -(K_inv_y .* da)' * K_inv_y + diag_K_inv' * da;
```

this corresponding to equation 29 in the Garnett (2017). 

The issue is `lya_1pz` now spans between 910 to 1600 A, so exp(-tau (1 + z)^beta) should be one if lambda > 1215.67

The code used in multi-DLA applied indicator on the `lya_1pz`. It was fine since there's no zero in `lya_1pz` in the multi-DLA code; however, we need to consider zeros in the z code.

The workaround is to inject the indicator after computing the gradient instead of directly in `lya_1pz`.

```matlab
 % Apr 12: inject indicator in the gradient but outside the log 
  indicator = lya_1pz <= zqso_1pz;
  da = da .* log(lya_1pz) * beta .* indicator;
  dlog_beta  = -(K_inv_y .* da)' * K_inv_y + diag_K_inv' * da;
```

Other changes are:
- change the initial beta, tau, and c_0 to the optimal values in multi-DLA paper; this is done in both `set_parameters.m` and `objective.m`.
- pass z_qsos to `objective.m`
- remove indicator applied on `lya_1pzs` in `learn_qso_model.m`
- And some unnecessary white spaces

### CODE REVIEW

**Comment by jibanCat** [comment]
My learning script is running now, though haven't finished, there's no stepsize invalid thing anymore.

Will report again here with the trained values once it finished running.

File: learn_qso_model.m [hunk 1]
```diff
@@ -185,17 +186,17 @@
...
 centered_rest_fluxes = bsxfun(@minus, rest_fluxes_div_exp1pz, mu);
-clear('rest_fluxes', 'rest_fluxes_div_exp1pz');
+clear('rest_fluxes');
 
 % get top-k PCA vectors to initialize M
 [coefficients, ~, latent] = ...
     pca(centered_rest_fluxes, ...
         'numcomponents', k, ...
-        'rows',          'pairwise');
+        'rows',          'complete');
```
Comment (sbird) [inline learn_qso_model.m]: Could you revert this?

**Review by sbird — COMMENTED** [review]
Thanks Ming-Feng! I appreciate it. Did you mean to change the PCA back from pairwise to complete?

File: learn_qso_model.m (hunk 1 above)
Comment (jibanCat) [reply learn_qso_model.m]: Yes I can revert it.

I think both pairwise and complete are fine. 

The problem is in the `spectrum_loss.m`

Comment (sbird) [reply learn_qso_model.m]: Yup, I agree! complete sometimes fails if you send it particularly bitty data.

Comment (jibanCat) [reply learn_qso_model.m]: I am using complete now and it runs fine

I remember complete assume there no missing data in the training set, which actually is not true for our data which have many NaNs.

Comment (sbird) [reply learn_qso_model.m]: That's right - so pairwise is more appropriate, I think, but pairwise sometimes fails to return a positive definite result. As long as pairwise succeeds I think we prefer it (although it isn't really important)

**Comment by jibanCat** [comment]
my training result:

```
c_0    = 0.1455
beta   = 3.1890
tau_0  = 0.0068
```
which is very different from the multi-dla model but closer to the Kim prior.

**Pull request merged**